### ISSUE TITLE
Retire plugin/autopath

### ISSUE THREAD

**Issue by miekg** [coredns/coredns#4187]
# Should we retire *autopath*?

First: I thing *autopath* is a really nice feature and doing stuff on the server of which you know the client will do in the future is just plain awesome.

There are however really hard to fix bugs with the current setup. Not so much that *autopath* is wrong, but in k8s (it's only used there) we don't have all the info, so something it does the wrong thing.

This is hard to debug and also hard to fix (because k8s reasons). 

For more background see https://github.com/coredns/coredns/issues/3765#issuecomment-606790416 (and the thread) and https://github.com/coredns/coredns/issues/3765#issuecomment-606060002

So we can
1. Tune down what autopath does in functionality to make it always work (not sure if possible)
2. Fix the actual bug (hard, see above)
3. Drop autopath (easy, but we may not want to)

What do others think?

**Comment by chrisohaver** [comment]
No, IMO #3765 was caused by the tombstone deletion bugs that we fixed in 1.7.0.  Autopath was affected by that bug because it relies on a pod watch, and if the pod cache was out of date in one instance and not another (which is how the tombstone deletion bug would manifest), a client could get conflicting answers.

**Comment by chrisohaver** [comment]
Well - let me rephrase.  I don't think we should retire autopath because of #3765 (because I think we already fixed the root cause of that).

But we can consider retiring it for other reasons:
1. autopath+kubernetes has some of the same goals as nodelocaldns, so it could be considered an alternative.  Still, nodelocaldns does not in itself reduce the number of queries made from a client (it caches the answers on the local node so they are not as expensive).
2. nodelocaldns deployments are pretty common.  e.g. kubespray deploys it by default.
3. there is an active KEP for a nodelocaldns version of autopath, which may be arriving soonish (e.g. k8s 1.21). When this arrives, we should probably at least drop the k8s integration of autopath (I recall an interesting non-k8s use case a while ago, so perhaps not retire autopath completely).

**Comment by chrisohaver** [comment]
FYI, The non-k8s autopath use case I mentioned above: #1394

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Retire plugin..." ]
>Well - let me rephrase. I don't think we should retire autopath because of #
>3765 (because I think we already fixed the root cause of that).
>
>But we can consider retiring it for other reasons:
>
> 1. autopath+kubernetes has some of the same goals as nodelocaldns, so it could
>    be considered an alternative. Still, nodelocaldns does not in itself reduce
>    the number of queries made from a client (it caches the answers on the
>    local node so they are not as expensive).
> 2. nodelocaldns deployments are pretty common. e.g. kubespray deploys it by
>    default.
> 3. there is an active KEP for a nodelocaldns version of autopath, which may be
>    arriving soonish (e.g. k8s 1.21). When this arrives, we should probably at
>    least drop the k8s integration of autopath (I recall an interesting non-k8s
>    use case a while ago, so perhaps not retire autopath completely).

thanks for adding some color to this. I think autopath is cool enough to leave around, but
we should find an in-tree use for it.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Retire plugin..." ]
>FYI, The non-k8s autopath use case I mentioned above: #1394

ack, that looks good enough to keep it.

Ok, let's see how this pans out and then (prolly) remove the kubernetes support.

So:
TL:DR:

* keep for now
* once nodelocaldns is smarter revisit kubernetes support

Thanks!

/close

**Comment by chrisohaver** [comment]
> ack, that looks good enough to keep it.

I'll see if I can craft an example to the autopath README for that use case.
